Added gzipTileset script to tools

lilleyse opened the pull request:
* [x] Write tests
* [x] Promisify the module

pjcozzi commented:
Also:

* [ ] Update README.md with instructions.

lilleyse commented:
This is updated.

lilleyse commented:
Ok this is actually ready now.

pjcozzi commented on tools/README.md:
```diff
@@ -1 +1,20 @@
+
+### gzipTileset
+
+Detects whether the tileset is compressed or not and does the opposite.
```
Are you sure this is the right form factor, it sounds awkward.

* I thought we were going to have one processing pipeline and be able to build a series of stages in exactly the same way as we plan to do so for gltf-pipeline.  Instead, this looks like the plan is to make lots of separate tools.  Are we sure that is the right approach?  If so, are we also taking the same approach with gltf-pipeline?  If so, what about the ability to provide a single JSON config to create the pipeline stages?
* Why not have separate zip / unzip arguments instead of "doing the opposite."  Did you look at other command-line zip tools?


> Detects whether the tileset

Is this precise?  I would expect it to test each tile so, for example, if 3 of 10 tests are already zipped, it only zips the other 7.

lilleyse replied:
> I thought we were going to have one processing pipeline and be able to build a series of stages in exactly the same way as we plan to do so for gltf-pipeline.

Ah yeah, I wasn't really thinking about that when making this script but I'll start to support a pipeline approach. gltf-pipeline is about half-way there in terms of the grand vision.

> Why not have separate zip / unzip arguments instead of "doing the opposite." Did you look at other command-line zip tools?

I can do that.

> Is this precise? I would expect it to test each tile so, for example, if 3 of 10 tests are already zipped, it only zips the other 7.

I didn't really expect this to happen in practice, but it shouldn't be hard to support.

pjcozzi replied:
The other option - that I would be OK with - is that each stage can be a separate command-line program and then the master pipeline command-line program would take the JSON config; however, this is different than, for example, `git`, `gzip`, etc.

pjcozzi commented:
@lilleyse is this ready now?

lilleyse commented:
Not yet, I still need to build out the command line tool/json input.

lilleyse commented:
Ok this is updated now. Let me know what you think about the pipeline JSON format.

pjcozzi commented:
> gunzip

ungzip?

pjcozzi commented:
> node ./bin/3d-tiles-tools.js gzip -i ./specs/data/TilesetOfTilesets/ -o ./output/TilesetOfTilesets-gzipped/

* Can `-i` be implicit or explicit?
* Can `-o` be optional?

Basically can I run this?

```
node ./bin/3d-tiles-tools.js gzip ./specs/data/TilesetOfTilesets/
```

pjcozzi commented:
Can you setup travis to run jsHint and the unit tests and coveralls for test coverage?  See [gltf-pipeline](https://github.com/AnalyticalGraphicsInc/gltf-pipeline) for a simple example, but probably only 90% right with best practices.  Let me know if you need me to login to enable things.

pjcozzi commented:
Add instructions for running the tests and other npm scripts to the README.md.  You can somewhat copy from our other repos.

Basically, I want to get this repo fully setup for development while the code itself is so small and simple so we can easily onramp CIS 399 students in a few weeks.

pjcozzi commented:
The README.md is current written for advanced users.  Change the order of the presentation so the individual tools - and their options - are introduced first (e.g., so if I come to the repo asking "how do I gzip this tileset?" I can quickly find the easy way to do it), and then the pipeline is introduced for advanced users.

Also, in the next PR when the pipeline gets a second stage, please add a pipeline diagram so it is super clear how things work.

pjcozzi commented:
This is minor, but I wonder if there is any way to add a package.json to the root directory so that if someone runs `npm install` on the root directory like I did, we can display a message that says to run it on one of the sub-directories.  In general, you might want to look a bit at how repos with multiple node modules are designed.  I know it is common, but I don't know the best practices.

pjcozzi commented:
> Add instructions for running the tests and other npm scripts to the README.md. You can somewhat copy from our other repos.

For repoing coverage, I think you can exclude the `bin` directory.  See if we did so in `gltf-pipeline`.

pjcozzi commented:
There's a JSHint warning:
```
bin\3d-tiles-tools.js
  line 90  col 48  ['tilesOnly'] is better written in dot notation.

  ‼  1 warning
```

pjcozzi commented:
> Change the order of the presentation so the individual tools - and their options - are introduced first

Likewise, the pipeline JSON section should show all of each stages options perhaps using formatting similar to the glTF spec.  The glTF spec is actually generated from the JSON schema, but at this point, this does not need one.

pjcozzi commented:
> node ./bin/3d-tiles-tools.js pipeline -i ./specs/data/pipeline.json

Can we make this as simple as
```
node ./bin/3d-tiles-tools.js ./specs/data/pipeline.json
```

pjcozzi commented on tools/README.md:
````diff
@@ -1 +1,68 @@
+```json
+{
+    "input": "Tileset/",
+    "output": "TilesetGzipped/",
````
Is `output` optional?  If not, can it be?

pjcozzi commented on tools/index.js:
```diff
@@ -1,2 +1,4 @@
 module.exports = {
+    gzipTileset : require('./lib/gzipTileset'),
```
Perhaps just submit an issue for now to generate doc for these.

pjcozzi commented on tools/lib/gzipTileset.js:
```diff
@@ -0,0 +1,110 @@
+/**
+ * gzips or gunzips the input tileset.
+ *
+ * @param {String} inputDirectory Path to the input directory.
```
Consider putting this all in the `options` object (like Cesium) or maybe strictly follow a pattern where required parameters are not in `options` and required ones are.  Ask I mentioned above, I think `outputDirectory` should be optional.

pjcozzi commented on tools/lib/gzipTileset.js:
```diff
@@ -0,0 +1,110 @@
+ * @param {String} inputDirectory Path to the input directory.
+ * @param {Object} [outputDirectory] Path to the output directory.
+ * @param {Object} [options] Object with the following properties:
+ * @param {Boolean} [options.gzip=true] Whether to gzip or gunzip the tileset.
```
Should this have `options.overwrite=false` (exposed throughout, of course) so users don't accidentally overwrite with their output?

lilleyse replied:
In my experience it's always a hassle to remember to set overwrite to true if the default is false. Maybe we can give the option but set default to true? Is that even worth doing?

pjcozzi replied:
Just follow whatever the Unix philosophy is.

lilleyse replied:
Many unix command line tools do ask for permission before overwriting. Node libraries on the other hand like `fs` always overwrite. If I were to chose I prefer the overwrite always way.

pjcozzi replied:
What about Node cli apps?  It's fine for the library to do one thing and the cli to do the other, I suppose, let's just make sure we are doing what is standard.

pjcozzi commented on tools/lib/gzipTileset.js:
```diff
@@ -0,0 +1,110 @@
+           extension === '.vctr';
+}
+
+function isGzipped(file) {
```
This could be in its own `@private` file instead of duplicating it in the tests.

pjcozzi commented on tools/lib/gzipTileset.js:
```diff
@@ -0,0 +1,110 @@
+                                });
+
+                        });
+                }, {concurrency: 1024})
```
Are you sure about 1024?

Should this be an option?  I don't want to over optionify things, but if there isn't a one size fits all when this is integrated into a more complex pipeline and server, this should probably be configurable.

lilleyse replied:
I didn't give much thought to it, I suppose keeping the number like 1024 will prevent too many promises from getting created, but still keep the operation fast. I'm not really sure what the best values are for a server or if it matters. @mramato thoughts?


and am not really sure of the implications of it for a server.

mramato replied:
For a cli program, 1024 is fine, but if you want to be conservative, a number like 100 should work fine too and I doubt you would see a performance difference since you are likely to starve the drive you're reading from in both cases.

This function would need a lot of work before I would consider using it in a server environment.  At a quick glance:

1. You wouldn't buffer the entire list of files in memory up front, that would cause problems for huge tilesets
2. Output would never go to disk, instead I would expect a callback to be raised for each tile as it is zipped and let the caller decide where the data gets written.
3. You wouldn't call `console.log` directly, logging/progress would be the of the callback in 1.

pjcozzi commented on tools/lib/runPipeline.js:
```diff
@@ -0,0 +1,112 @@
+    options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+    var verbose = defaultValue(options.verbose, false);
+
+    var workingDirectory1 = path.join(path.dirname(inputPath), path.basename(inputPath) + '-working1');
```
Here and below would be more robust if we use a working directory name that we know doesn't exist.

pjcozzi commented:
If this were ran on a server, are there any security concerns we need to address now, e.g., reading random directories, etc.

pjcozzi commented:
Just those comments, this project is off to a nice start!

lilleyse commented:
> Can you setup travis to run jsHint and the unit tests and coveralls for test coverage? See gltf-pipeline for a simple example, but probably only 90% right with best practices. Let me know if you need me to login to enable things.

You'll need to do this, I don't have admin access.

> This is minor, but I wonder if there is any way to add a package.json to the root directory so that if someone runs npm install on the root directory like I did, we can display a message that says to run it on one of the sub-directories. In general, you might want to look a bit at how repos with multiple node modules are designed. I know it is common, but I don't know the best practices.

I'm not sure if it's possible to override `npm install`.

> If this were ran on a server, are there any security concerns we need to address now, e.g., reading random directories, etc.

Yeah we would need to handle the input and output directories given in pipeline.json. I'll look into this some more.

> Output directory optional

Just replying to a lot of related comments here, the output directory is always optional.

lilleyse commented:
Made some changes, still have a few more things to round out.

lilleyse commented:
> If this were ran on a server, are there any security concerns we need to address now, e.g., reading random directories, etc.

I believe it should be safe as long as the server that wraps this project writes the input files and output files to temp directories before processing, as well as changes the relative paths in the pipeline JSON to absolute paths (like in `3d-tiles-tools.js`). I'm not sure if there's much we can do here.

pjcozzi commented:
>> Can you setup travis to run jsHint and the unit tests and coveralls for test coverage? See gltf-pipeline for a simple example, but probably only 90% right with best practices. Let me know if you need me to login to enable things.
> You'll need to do this, I don't have admin access.

I'll submit an issue; it will be a bit before I have time for this.

Anything else left for this PR?

lilleyse commented:
> Anything else left for this PR?

This should be ready now. `-f, --force` is now added to overwrite existing folders.

lilleyse commented:
The only unresolved comment is https://github.com/AnalyticalGraphicsInc/3d-tiles-tools/pull/3#discussion_r75106486.

mramato commented on tools/bin/3d-tiles-tools.js:
```diff
@@ -0,0 +1,96 @@
+
+if (process.argv.length < 4 || defined(argv.h) || defined(argv.help) || !defined(argv._[0])) {
+    var help =
+        'Usage: node ' + path.basename(__filename) + ' command [args]\n' +
```
`node` should not be here.  this repository should be installed with `-g` at which point every command would be in the user's system path.

mramato commented on tools/lib/gzipTileset.js:
```diff
@@ -0,0 +1,110 @@
+        if (!defined(inputDirectory)) {
+            reject(new DeveloperError('inputDirectory is required'));
+        }
+        inputDirectory = path.normalize(inputDirectory);
```
`path.normalize` for this and the next line really belong in the calling code, not here.

mramato commented on tools/lib/gzipTileset.js:
```diff
@@ -0,0 +1,110 @@
+        var verbose = defaultValue(options.verbose, false);
+
+        if (!defined(inputDirectory)) {
+            reject(new DeveloperError('inputDirectory is required'));
```
Because it's a `DeveloperError`, this check should happen outside of the Promise and throw immediately.

mramato commented:
> The only unresolved comment is #3 (comment).

Just to clarify, I don't think you necessarily needy to make any of those changes.  I was just trying to illustrate how writing server-side tools is very different than writing cli tools.  Of course if our goal is to provide a library of tools that will be used either on the client or the server, then we need to lay out some architecture ground rules up front.

pjcozzi commented:
>> The only unresolved comment is #3 (comment).
> Just to clarify, I don't think you necessarily needy to make any of those changes. I was just trying to illustrate how writing server-side tools is very different than writing cli tools. Of course if our goal is to provide a library of tools that will be used either on the client or the server, then we need to lay out some architecture ground rules up front.

If this is easy enough to address now, I would like to do it soon.

@mramato please keep a backlog of updates for the Node Guide.

lilleyse commented:
I addressed https://github.com/AnalyticalGraphicsInc/3d-tiles-tools/pull/3#r77263777.

Now the tiles are gzipped during the directory walk rather than all at once. To check that all the files are processed, I keep track of the current count and compare that with the total number of files in the directory.

I also added writing and logging callbacks. However I'm not completely sure what to do with the write callbacks when there are multiple stages. Currently intermediary stages ping-pong between two temp directories. Should the write callback have control over those too? I have a TODO in the code for this.

@mramato I would appreciate your thoughts on the current approaches.

pjcozzi commented:
@mramato if you have any more feedback, please submit an issue.
